etc/systemd/system-generators/zfs-mount-generator.in

```diff
@@ -151,12 +151,17 @@ ExecStop=@sbindir@/zfs unload-key '${dataset}'"   > "${dest_norm}/${keyloadunit}
     return
   elif [ "${p_mountpoint}" = "none" ] ; then
     return
-  elif [ "${p_mountpoint%"${p_mountpoint#?}"}" != "/" ] ; then
+  elif [ "${p_mountpoint%"${p_mountpoint#/}"}" != "/" ] ; then
     do_fail "invalid mountpoint $*"
   fi
 
   # Escape the mountpoint per systemd policy.
-  mountfile="$(systemd-escape "${p_mountpoint#?}").mount"
+  if [ "${p_mountpoint}" = "/" ] ; then
+    mountfile=-.mount
```

Comment:
Could you site `systemd`'s `src/basic/unit-name.c:unit_name_path_escape` function here? Or, if you are aware of it, an official statement of this behavior?

Reply:
I believe the systemd.unit(5) man page has the relevant bits:
```
STRING ESCAPING FOR INCLUSION IN UNIT NAMES
[...]
The escaping algorithm operates as follows: given a string, any "/" character is
replaced by "-", and all other characters which are not ASCII alphanumerics or "_"
are replaced by C-style "\x2d" escapes. In addition, "."  is replaced with such a
C-style escape when it would appear as the first character in the escaped string.

When the input qualifies as absolute file system path, this algorithm is extended
slightly: the path to the root directory "/" is encoded as single dash "-". In
addition, any leading, trailing or duplicate "/" characters are removed from the
string before transformation. Example: /foo//bar/baz/ becomes "foo-bar-baz".
```

but the next paragraph also tells us we could do away with the stripping and special casing:

```
This escaping is fully reversible, as long as it is known whether the escaped string
was a path (the unescaping results are different for paths and non-path strings). The
systemd-escape(1) command may be used to apply and reverse escaping on arbitrary
strings. Use systemd-escape --path to escape path strings, and systemd-escape without
--path otherwise.
```

Reply:
That settles it: we should replace all this logic here with `systemd-escape --path --suffix=mount ${p_mountpoint}`